lib/User/Manager.php

```diff
@@ -294,32 +295,71 @@ public function resolveUID(UserEntry $userEntry) {
 
 		$dn = $userEntry->getDN();
 
-		//let's try to retrieve the ownCloud name from the mappings table
-		$ocName = trim($mapper->getNameByDN($dn));
-		if(is_string($ocName) && $ocName !== '') {
+		// let's try to retrieve the ownCloud name from the mappings table
+		$ocName = \trim($mapper->getNameByDN($dn));
+		if(\is_string($ocName) && $ocName !== '') {
 			return $ocName;
 		}
 
-		//second try: get the UUID and check if it is known. Then, update the DN and return the name.
+		// second try: get the UUID and check if it is known. Then, update the DN and return the name.
 		$uuid = $userEntry->getUUID();
-		$ocName = trim($mapper->getNameByUUID($uuid));
-		if(is_string($ocName) && $ocName !== '') {
+		$ocName = \trim($mapper->getNameByUUID($uuid));
+		if(\is_string($ocName) && $ocName !== '') {
+			$this->logger->warning(
+				'DN changed! Found user {name} by uuid {uuid}, updating dn to {dn}',
+				[
+					'app'  => __METHOD__,
+					'name' => $ocName,
+					'uuid' => $uuid,
+					'dn'   => $dn
+				]
+			);
 			$mapper->setDNbyUUID($dn, $uuid);
 			return $ocName;
 		}
 
-		$intName = trim($this->access->sanitizeUsername($userEntry->getUsername()));
+
+		// third try: fallback to poorly escaped dn
+		if((bool)$this->ocConfig->getAppValue('user_ldap', 'resolve_uid_by_legacy_dn', true)) {
```

Comment:
I'd prefer to keep the consistency with other similar configuration checks. I'm not sure if this will work properly.
`$this->ocConfig->getAppValue('user_ldap', 'resolve_uid_by_legacy_dn', 'true') === 'true'`

Reply:
Is it worthy to write a migration to fix this case?

Reply:
I'll use `'yes'` and `'no'`, most places use that:
```
grep -R getAppValue * | grep '=='
apps/files_external/lib/Panels/Admin.php:		$tmpl->assign('enableExternalStorage', $this->config->getAppValue('core', 'enable_external_storage', 'no') === 'yes');
apps/files_external/lib/Panels/Admin.php:		$tmpl->assign('allowUserMountSharing', $this->config->getAppValue('core', 'allow_user_mount_sharing', 'yes') === 'yes');
apps/files_external/lib/Panels/Personal.php:		$enabled = ($this->config->getAppValue('core', 'enable_external_storage', 'no') === 'yes'
apps/files_external/lib/Panels/Personal.php:		$tmpl->assign('allowUserMountSharing', $this->config->getAppValue('core', 'allow_user_mount_sharing', 'yes') === 'yes');
apps/files_external/appinfo/app.php:if ($config->getAppValue('core', 'enable_external_storage', 'no') === 'yes') {
apps/files_sharing/lib/Controller/ShareesController.php:		$this->shareWithGroupOnly = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members', 'no') === 'yes';
apps/files_sharing/lib/Controller/ShareesController.php:		$this->shareWithMembershipGroupOnly = $this->config->getAppValue('core', 'shareapi_only_share_with_membership_groups', 'no') === 'yes';
apps/files_sharing/lib/Controller/ShareesController.php:		$this->shareeEnumeration = $this->config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'yes') === 'yes';
apps/files_sharing/lib/Controller/ShareesController.php:			$this->shareeEnumerationGroupMembers = $this->config->getAppValue('core', 'shareapi_share_dialog_user_enumeration_group_members', 'no') === 'yes';
apps/files_sharing/lib/Middleware/SharingCheckMiddleware.php:			$this->config->getAppValue('files_sharing', 'incoming_server2server_share_enabled', 'yes') !== 'yes') {
apps/files_sharing/lib/Middleware/SharingCheckMiddleware.php:			$this->config->getAppValue('files_sharing', 'outgoing_server2server_share_enabled', 'yes') !== 'yes') {
apps/files_sharing/lib/Middleware/SharingCheckMiddleware.php:		if ($this->config->getAppValue('core', 'shareapi_enabled', 'yes') !== 'yes') {
apps/files_sharing/lib/Middleware/SharingCheckMiddleware.php:		if ($this->config->getAppValue('core', 'shareapi_allow_links', 'yes') !== 'yes') {
apps/files_sharing/lib/API/Share20OCS.php:		$autoAccept = $this->config->getAppValue('core', 'shareapi_auto_accept_share', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:		if ($this->config->getAppValue('core', 'shareapi_enabled', 'yes') !== 'yes') {
apps/files_sharing/lib/Capabilities.php:			$public['enabled'] = $this->config->getAppValue('core', 'shareapi_allow_links', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$roPasswordEnforced = $this->config->getAppValue('core', 'shareapi_enforce_links_password_read_only', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$rwPasswordEnforced = $this->config->getAppValue('core', 'shareapi_enforce_links_password_read_write', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$woPasswordEnforced = $this->config->getAppValue('core', 'shareapi_enforce_links_password_write_only', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$public['expire_date']['enabled'] = $this->config->getAppValue('core', 'shareapi_default_expire_date', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:					$public['expire_date']['enforced'] = $this->config->getAppValue('core', 'shareapi_enforce_expire_date', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$public['send_mail'] = $this->config->getAppValue('core', 'shareapi_allow_public_notification', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$public['social_share'] = $this->config->getAppValue('core', 'shareapi_allow_social_share', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$public['upload'] = $this->config->getAppValue('core', 'shareapi_allow_public_upload', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:			$res['user']['send_mail'] = $this->config->getAppValue('core', 'shareapi_allow_mail_notification', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:			$res['resharing'] = $this->config->getAppValue('core', 'shareapi_allow_resharing', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:			$res['group_sharing'] = $this->config->getAppValue('core', 'shareapi_allow_group_sharing', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:			$res['auto_accept_share'] = $this->config->getAppValue('core', 'shareapi_auto_accept_share', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:			$res['share_with_group_members_only'] = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:			$res['share_with_membership_groups_only'] = $this->config->getAppValue('core', 'shareapi_only_share_with_membership_groups', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:			$user_enumeration['enabled'] = $this->config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'yes') === 'yes';
apps/files_sharing/lib/Capabilities.php:				$user_enumeration['group_members_only'] = $this->config->getAppValue('core', 'shareapi_share_dialog_user_enumeration_group_members', 'no') === 'yes';
apps/files_sharing/lib/Capabilities.php:			'outgoing'  => $this->config->getAppValue('files_sharing', 'outgoing_server2server_share_enabled', 'yes') === 'yes',
apps/files_sharing/lib/Capabilities.php:			'incoming' => $this->config->getAppValue('files_sharing', 'incoming_server2server_share_enabled', 'yes') === 'yes'
apps/files_sharing/appinfo/app.php:		if (\OC::$server->getConfig()->getAppValue('files_sharing', 'incoming_server2server_share_enabled', 'yes') === 'yes') {
apps/files_sharing/appinfo/app.php:if (\class_exists('OCA\Files\App') && $config->getAppValue('core', 'shareapi_enabled', 'yes') === 'yes') {
apps/files_sharing/appinfo/app.php:		if ($config->getAppValue('core', 'shareapi_allow_links', 'yes') === 'yes') {
apps/testing/lib/Application.php:		if ($config->getAppValue($appName, 'enable_alt_user_backend', 'no') === 'yes') {
apps/testing/lib/Application.php:			if ($config->getAppValue($appName, 'updated_all', 'no') === 'no') {
apps-repos/user_cas/lib/Service/AppService.php:        if ($this->config->getAppValue($this->appName, 'cas_force_login') !== '1') {
apps-repos/encryption/lib/Command/SelectEncryptionType.php:		if ($this->config->getAppValue('core', 'encryption_enabled', 'no') === 'no') {
apps-repos/encryption/lib/Command/SelectEncryptionType.php:		$userKeysNotEnabled = ($this->config->getAppValue('encryption', 'userSpecificKey', '') === '');
apps-repos/encryption/lib/KeyManager.php:		if($this->config->getAppValue('encryption', 'masterKeyId') !== $this->masterKeyId) {
grep: build/bin/bower: Datei oder Verzeichnis nicht gefunden
grep: custom: Datei oder Verzeichnis nicht gefunden
core/js/config.php:$enforceLinkPasswordReadOnly = $config->getAppValue('core', 'shareapi_enforce_links_password_read_only', 'no') === 'yes';
core/js/config.php:$enforceLinkPasswordReadWrite = $config->getAppValue('core', 'shareapi_enforce_links_password_read_write', 'no') === 'yes';
core/js/config.php:$enforceLinkPasswordWriteOnly = $config->getAppValue('core', 'shareapi_enforce_links_password_write_only', 'no') === 'yes';
core/js/config.php:$outgoingServer2serverShareEnabled = $config->getAppValue('files_sharing', 'outgoing_server2server_share_enabled', 'yes') === 'yes';
core/Migrations/Version20170221121536.php:		$filesExternalEnabled = $config->getAppValue('files_external', 'enabled', 'no') === 'yes';
core/ajax/share.php:					->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'yes') === 'yes';
core/Command/Encryption/Enable.php:		if ($this->config->getAppValue('core', 'encryption_enabled', 'no') === 'yes') {
core/Command/Encryption/Disable.php:		if ($this->config->getAppValue('core', 'encryption_enabled', 'no') !== 'yes') {
lib/private/Installer.php:		return (\OC::$server->getConfig()->getAppValue($app, "installed_version", null) !== null);
lib/private/Installer.php:									&& \OC::$server->getConfig()->getAppValue($filename, 'enabled') !== 'no') {
lib/private/legacy/template.php:				if (\OC::$server->getConfig()->getAppValue('core', 'backgroundjobs_mode', 'ajax') == 'ajax') {
lib/private/legacy/app.php:		} elseif (\OC::$server->getConfig()->getAppValue($appId, 'ocsid', null) !== null) {
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_enabled', 'yes') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_allow_links', 'yes') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_enforce_links_password_read_only', 'no') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_enforce_links_password_read_write', 'no') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_enforce_links_password_write_only', 'no') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_default_expire_date', 'no') === 'yes';
lib/private/Share20/Manager.php:			$this->config->getAppValue('core', 'shareapi_enforce_expire_date', 'no') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_allow_public_upload', 'yes') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_only_share_with_group_members', 'no') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_only_share_with_membership_groups', 'no') === 'yes';
lib/private/Share20/Manager.php:		return $this->config->getAppValue('core', 'shareapi_allow_group_sharing', 'yes') === 'yes';
lib/private/Share20/Manager.php:		if ($this->config->getAppValue('core', 'shareapi_exclude_groups', 'no') === 'yes') {
lib/private/Share20/Manager.php:		return $this->config->getAppValue('files_sharing', 'outgoing_server2server_share_enabled', 'yes') === 'yes';
lib/private/Updater/VersionCheck.php:		if ($this->config->getAppValue('core', 'installedat', '') === '') {
lib/private/Files/External/ConfigAdapter.php:		$allowUserMountSharing = $this->config->getAppValue('core', 'allow_user_mount_sharing', 'yes') === 'yes';
lib/private/Files/External/StoragesBackendService.php:		if ($this->config->getAppValue('files_external', 'allow_user_mounting', 'yes') !== 'yes') {
lib/private/Server.php:			if ($config->getAppValue('core', 'enable_external_storage', 'no') === 'yes') {
lib/private/Updater.php:		if ($this->getVendor() !== $this->config->getAppValue('core', 'vendor', '')) {
settings/Controller/CheckSetupController.php:		if ($this->config->getAppValue('files_sharing', 'outgoing_server2server_share_enabled', 'yes') === 'no'
settings/Controller/CheckSetupController.php:			&& $this->config->getAppValue('files_sharing', 'incoming_server2server_share_enabled', 'yes') === 'no') {
settings/Panels/Admin/FileSharing.php:		$template->assign('onlyShareWithMembershipGroups', $this->config->getAppValue('core', 'shareapi_only_share_with_membership_groups', 'no') === 'yes');
settings/Panels/Admin/FileSharing.php:		$excludeGroups = $this->config->getAppValue('core', 'shareapi_exclude_groups', 'no') === 'yes' ? true : false;
settings/Panels/Admin/SecurityWarning.php:		$shouldSuggestOverwriteCliUrl = $this->config->getAppValue('core', 'backgroundjobs_mode', 'ajax') === 'cron' &&
```

Reply:
proper way to check the variable seems to be `filter_var($allowEnum, FILTER_VALIDATE_BOOLEAN)`, see https://stackoverflow.com/a/15075609. I'll use that everywhere in user_ldap to make it more robust.